tools/src/main/java/org/apache/kafka/tools/ConsumerPerformance.java

```diff
@@ -0,0 +1,394 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.tools;
+
+import joptsimple.OptionException;
+import joptsimple.OptionSpec;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.utils.Exit;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.server.util.CommandDefaultOptions;
+import org.apache.kafka.server.util.CommandLineUtils;
+import org.apache.kafka.server.util.ToolsUtils;
+
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class ConsumerPerformance {
+    private static final Random RND = new Random();
+
+    public static void main(String[] args) {
+        try {
+            ConsumerPerfOptions options = new ConsumerPerfOptions(args);
+            // group counters
+            AtomicLong totalMessagesRead = new AtomicLong(0), totalBytesRead = new AtomicLong(0),
+                joinGroupTimeMs = new AtomicLong(0);
+            // consumer counters
+            long bytesRead = 0L, messagesRead = 0L, lastBytesRead = 0L, lastMessagesRead = 0L,
+                joinTimeMsInSingleRound = 0L;
+
+            if (!options.hideHeader())
+                printHeader(options.showDetailedStats());
+
+            KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(options.props());
```

Comment:
Maybe create via instance method to add a unit test which exercises the consume on a mock?

Reply:
Thought about this, but in the end I decided to migrate the test as is because it is similar to other tools. We can always do some refactoring later (e.g. command interface).

Reply:
Makes sense.